pennylane/devices/qubit/preprocess.py

```diff
@@ -95,34 +100,18 @@ def validate_and_expand_adjoint(
         Union[.QuantumTape, .DeviceError]: The expanded tape, such that it is supported by adjoint differentiation.
         If the circuit is invalid for adjoint differentiation, a DeviceError with an explanation is returned instead.
     """
-    # Check validity of measurements
-    measurements = []
-    for m in circuit.measurements:
-        if not isinstance(m, ExpectationMP):
-            return DeviceError(
-                "Adjoint differentiation method does not support "
-                f"measurement {m.__class__.__name__}."
-            )
-
-        if not m.obs.has_matrix:
-            return DeviceError(
-                f"Adjoint differentiation method does not support observable {m.obs.name}."
-            )
-
-        measurements.append(m)
 
-    expanded_ops = []
-    for op in circuit._ops:
-        if op.num_params > 1:
-            if not isinstance(op, qml.Rot):
-                return DeviceError(
-                    f"The {op} operation is not supported using "
-                    'the "adjoint" differentiation method.'
-                )
-            ops = op.decomposition()
-            expanded_ops.extend(ops)
-        elif not isinstance(op, qml.operation.StatePrep):
-            expanded_ops.append(op)
+    try:
+        new_ops = [
+            final_op
+            for op in circuit._ops
+            for final_op in _operator_decomposition_gen(op, _accepted_adjoint_operator)
+        ]
+    except RecursionError as e:
+        raise DeviceError(
+            "Reached recursion limit trying to decompose operations. "
+            "Operator decomposition may have entered an infinite loop."
+        ) from e
```

Comment:
is it worth putting this subroutine into its own function, since it's a copy-paste from expand_fn?

Reply:
Probably. We can probably come back through and tidy all this up at some point.